test/object-store/sync/flx_role_change.cpp

```diff
@@ -0,0 +1,861 @@
+
+#include <realm/util/future.hpp>
+#include <realm/util/logger.hpp>
+
+#include <filesystem>
+#include <iostream>
+#include <stdexcept>
+
+using namespace realm;
+using namespace realm::app;
+
+namespace {
+
+const Schema g_person_schema{{"Person",
+                              {{"_id", PropertyType::ObjectId, Property::IsPrimary{true}},
+                               {"role", PropertyType::String},
+                               {"name", PropertyType::String},
+                               {"emp_id", PropertyType::Int}}}};
+
+auto fill_person_schema = [](SharedRealm realm, std::string role, size_t count) {
+    CppContext c(realm);
+    for (size_t i = 0; i < count; ++i) {
+        auto obj = Object::create(c, realm, "Person",
+                                  std::any(AnyDict{
+                                      {"_id", ObjectId::gen()},
+                                      {"role", role},
+                                      {"name", util::format("%1-%2", role, i)},
+                                      {"emp_id", static_cast<int64_t>(i)},
+                                  }));
+    }
+};
+
+struct TestParams {
+    size_t num_emps = 150;
+    size_t num_mgrs = 25;
+    size_t num_dirs = 10;
+    std::optional<size_t> num_objects = 10;
+    std::optional<size_t> max_download_bytes = 4096;
+    std::optional<size_t> sleep_millis = std::nullopt;
+};
+
+std::unique_ptr<FLXSyncTestHarness> setup_harness(std::string app_name, TestParams params)
+{
+    auto harness = std::make_unique<FLXSyncTestHarness>(
+        app_name, FLXSyncTestHarness::ServerSchema{g_person_schema, {"role", "name"}});
+
+    auto& app_session = harness->session().app_session();
+
+    if (params.num_objects) {
+        REQUIRE(app_session.admin_api.patch_app_settings(
+            app_session.server_app_id, {{"sync", {{"num_objects_before_bootstrap_flush", *params.num_objects}}}}));
+    }
+
+    if (params.max_download_bytes) {
+        REQUIRE(app_session.admin_api.patch_app_settings(
+            app_session.server_app_id,
+            {{"sync", {{"qbs_download_changeset_soft_max_byte_size", *params.max_download_bytes}}}}));
+    }
+
+    if (params.sleep_millis) {
+        REQUIRE(app_session.admin_api.patch_app_settings(
+            app_session.server_app_id, {{"sync", {{"download_loop_sleep_millis", *params.sleep_millis}}}}));
+    }
+
+    // Initialize the realm with some data
+    harness->load_initial_data([&](SharedRealm realm) {
+        fill_person_schema(realm, "employee", params.num_emps);
+        fill_person_schema(realm, "manager", params.num_mgrs);
+        fill_person_schema(realm, "director", params.num_dirs);
+    });
+    // Return the unique_ptr for the newly created harness
+    return harness;
+}
+
+void teardown_harness(std::unique_ptr<FLXSyncTestHarness>& harness)
+{
+    if (!harness) {
+        return; // nothing to do
+    }
+    harness->app()->sync_manager()->wait_for_sessions_to_terminate();
+    harness.reset();
```

Comment:
Do we re-use the harness unique_ptr? Why not just take the unique_ptr by value and let it go out of scope? Why have this function at all?

Reply:
A single harness is re-used across the sections within a test_case. I ended up removing this function and just call `harness.reset()` in the last section.